unattended-upgrade

```diff
@@ -155,55 +155,61 @@ class UnattendedUpgradesCache(apt.Cache):
-            allowed origin
+    def adjust_candidate(self, pkg):
+        # type: (apt.Package) -> Bool
+        """ Adjust origin and return True if adjustment took place
 
-            This lets adjusting the origin even if the candidate has a higher
-            version. This is needed when e.g. a package is available in
+            This is needed when e.g. a package is available in
             the security pocket but there is also a package in the
             updates pocket with a higher version number
         """
-        candidates = {}
+        new_cand = None
+        for ver in pkg.versions:
+            # ignore versions that the user marked with priority < 100
+            # (and ensure we have a python-apt that supports this)
+            if (hasattr(ver, "policy_priority") and ver.policy_priority < 100):
+                logging.debug("ignoring ver %s with priority < 0" % ver)
+                continue
+            if is_allowed_origin(ver, self.allowed_origins):
+                # leave as soon as we have the highest new candidate
+                new_cand = ver
+                break
+        if new_cand and new_cand != pkg.candidate:
+            logging.debug("adjusting candidate version: %s" % new_cand)
+            pkg.candidate = new_cand
+            return True
+        return False
+
+    def adjust_candidates(self):
+        # tyoe: () -> None
+        """ Adjust all package candidates when needed saving cached package
+            names to self._cached_candidate_pkgnames
+        """
         if self._cached_candidate_pkgnames is None:
-            pkg_iter = iter(self.keys())
+            # first run, collect package names
+            self._cached_candidate_pkgnames = set()
+            for pkg in self:
+                # important! this avoids downgrades below
+                if pkg.is_installed and not pkg.is_upgradable:
+                    continue
+                # check if the candidate is already pointing to a allowed
+                # origin and if so, do not mess with it
+                if is_allowed_origin(pkg.candidate, self.allowed_origins):
+                    continue
+                # check if we have a version in a allowed origin that is
+                # not the candidate
+                if self.adjust_candidate(pkg):
+                    # ... and save it for next runs
+                    self._cached_candidate_pkgnames.add(pkg.name)
         else:
-            pkg_iter = iter(
-                [self[name] for name in self._cached_candidate_pkgnames])
-        for pkgname in pkg_iter:
-            pkg = self[pkgname]
-            # important! this avoids downgrades below
-            if pkg.is_installed and not pkg.is_upgradable:
-                continue
-            # check if the candidate is already pointing to a allowed
-            # origin and if so, do not mess with it
-            if is_allowed_origin(pkg.candidate, self.allowed_origins):
```

Comment:
In the original code this check made pkgs with already adjusted candidates being skipped when re-generating the candidates dictionary, effectively making u-u reseting _cached_candidate_pkgnames to {}.